In [216]:
import pandas as pd
import numpy as np

In [217]:
df = pd.read_csv('kc_house_data.csv')
df

id             date     price  bedrooms  bathrooms  \
0      7129300520  20141013T000000  221900.0         3       1.00   
1      6414100192  20141209T000000  538000.0         3       2.25   
2      5631500400  20150225T000000  180000.0         2       1.00   
3      2487200875  20141209T000000  604000.0         4       3.00   
4      1954400510  20150218T000000  510000.0         3       2.00   
...           ...              ...       ...       ...        ...   
21608   263000018  20140521T000000  360000.0         3       2.50   
21609  6600060120  20150223T000000  400000.0         4       2.50   
21610  1523300141  20140623T000000  402101.0         2       0.75   
21611   291310100  20150116T000000  400000.0         3       2.50   
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
0             1180      5650     1.0           0     0  ...      7   
1             2570      7242     2.0           0     0  ...      7   
2              770     10000     1.0           0     0  ...      6   
3             1960      5000     1.0           0     0  ...      7   
4             1680      8080     1.0           0     0  ...      8   
...            ...       ...     ...         ...   ...  ...    ...   
21608         1530      1131     3.0           0     0  ...      8   
21609         2310      5813     2.0           0     0  ...      8   
21610         1020      1350     2.0           0     0  ...      7   
21611         1600      2388     2.0           0     0  ...      8   
21612         1020      1076     2.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
0            1180              0      1955             0    98178  47.5112   
1            2170            400      1951          1991    98125  47.7210   
2             770              0      1933             0    98028  47.7379   
3            1050            910      1965             0    98136  47.5208   
4            1680              0      1987             0    98074  47.6168   
...           ...            ...       ...           ...      ...      ...   
21608        1530              0      2009             0    98103  47.6993   
21609        2310              0      2014             0    98146  47.5107   
21610        1020              0      2009             0    98144  47.5944   
21611        1600              0      2004             0    98027  47.5345   
21612        1020              0      2008             0    98144  47.5941   

          long  sqft_living15  sqft_lot15  
0     -122.257           1340        5650  
1     -122.319           1690        7639  
2     -122.233           2720        8062  
3     -122.393           1360        5000  
4     -122.045           1800        7503  
...        ...            ...         ...  
21608 -122.346           1530        1509  
21609 -122.362           1830        7200  
21610 -122.299           1020        2007  
21611 -122.069           1410        1287  
21612 -122.299           1020        1357  

[21613 rows x 21 columns]

In [218]:
df.columns

Index(['id', 'date', 'price', 'bedrooms', 'bathrooms', 'sqft_living',
       'sqft_lot', 'floors', 'waterfront', 'view', 'condition', 'grade',
       'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode',
       'lat', 'long', 'sqft_living15', 'sqft_lot15'],
      dtype='object')

At first look, the dataset we have describes a group of factors for thousands of homes, most importantly the price column, which will be our main focus and goal to find its correlation with other factors and how does it change

In [219]:
for col in df.columns:
    temp = df[col].unique()
    if len(temp) > 10:
        temp = len(temp)
    print(f'{col}: {temp}')

id: 21436
date: 372
price: 3625
bedrooms: 13
bathrooms: 30
sqft_living: 1038
sqft_lot: 9782
floors: [1.  2.  1.5 3.  2.5 3.5]
waterfront: [0 1]
view: [0 3 4 2 1]
condition: [3 5 4 1 2]
grade: 12
sqft_above: 946
sqft_basement: 306
yr_built: 116
yr_renovated: 70
zipcode: 70
lat: 5034
long: 752
sqft_living15: 777
sqft_lot15: 8689


so here we got unique values for each factor or number of unique value if they are a lot.  
In this way we can take a glance on our factors to understand their meaning  
* we have several quantitative factors related to house area, sqft_lot, sqft_living, sqft_above.. intereseted to see the difference btw them and their effect on price  
* several qualitative variables describe the house, floors, bedroom, bathrooms.. I assume they all impact house price positively
* number of unique values in zipcode column is not big, maybe the houses are not distributed on many different areas, we can later study that using also lat & long variables  
* yr_built 116 unique values indicates there is a big range, so houses were built over a long period not at the same time  
* waterfront has only 2 unique values (0,1) so it is a flag whether a house has a waterfront or no, which I feel is a factor that makes a big difference for a house

In [220]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

good thing is that our data has no missing values which is good

so now we can start our analysis, we will try to dive in each factor and get insights about it, relate it to price, and find -if possible and beneficial- relation with other factors too

### Waterfront

In [221]:
mean1 = df.loc[df.waterfront==1, 'price'].mean()
mean0 = df.loc[df.waterfront==0, 'price'].mean()
print(f"Houses with waterfront have a mean price of {mean1.round(-3)} greater by {((mean1-mean0)/mean0).round(4)*100}% than that of houses without waterfront which is {mean0.round(-3)}")

Houses with waterfront have a mean price of 1663000.0 greater by 212.71% than that of houses without waterfront which is 532000.0


having a waterfront is big advantage, with triple price than others, but what about its effect on other factors?

In [222]:
df_grouped_water = df[['id', 'price', 'grade', 'waterfront', 'view']].groupby('waterfront').agg({'id':'count', 'price':'mean', 'grade':'median', 'view':'median'}).reset_index()
df_grouped_water.rename(columns={'id':'count'}, inplace=True)
df_grouped_water['percentage'] = df_grouped_water.apply(lambda x:x['count']/df.shape[0] *100,axis=1)
df_grouped_water

waterfront  count         price  grade  view  percentage
0           0  21450  5.316534e+05    7.0   0.0   99.245824
1           1    163  1.662524e+06    9.0   4.0    0.754176

##### So houses with waterfront consequently have better view, higher grade, higher price. However, only little percentage of houses have waterfront

### Houses built after 2000

Logically speaking, a new house would be better than old one, but there are many factors that should be taken into consideration.  
We will make a flag column that indicates if a house is new or not

In [223]:
df['new'] = df.yr_built >= 2000

In [224]:
percentile = 60
print(f'To begin, {(df.loc[df.new].shape[0]/df.shape[0])*100:.2f}% of houses are built from 2000 onwards, so most houses are actually old, where {percentile}% of houses were built even before {int(np.percentile(df.yr_built, percentile))}\nOldest house was built in {df.yr_built.min()}, newest in {df.yr_built.max()}')

To begin, 22.02% of houses are built from 2000 onwards, so most houses are actually old, where 60% of houses were built even before 1983
Oldest house was built in 1900, newest in 2015


In [225]:
p_after = df[df.new].price.mean()
p_before = df[~df.new].price.mean()
print(f'Houses built from 2000 onwards have a mean price approximate to {p_after.round(-3)} larger by {(((p_after - p_before)/p_before)*100).round(2)}% than that of houses before 2000 which is about {p_before.round(-3)}')

Houses built from 2000 onwards have a mean price approximate to 618000.0 larger by 19.38% than that of houses before 2000 which is about 518000.0


So yeah new houses have higher price, is it only because they are newer?

In [226]:
print(f'\nThere are {df[df.new].waterfront.sum()} new houses with watefronts, and {df[~df.new].waterfront.sum()} old houses with')


There are 20 new houses with watefronts, and 143 old houses with


this may seem to contradict with previous results, didn't we say that waterfront leverages the house price?  
but that knowing that houses with watefronts are rare, this difference proportionally to the total number of houses, does not make a difference

In [227]:
print(f'\n{(df[df.new].waterfront.sum()/df[df.new].waterfront.count() * 100).round(2)}% of new houses have watefronts, while {(df[~df.new].waterfront.sum()/df[~df.new].waterfront.count() * 100).round(2)}% of old houses have')


0.42% of new houses have watefronts, while 0.85% of old houses have


old houses have double percentage, but it is only 0.4% difference  
  
It is always important to check if the conculsions we've made are correct, and in this example we've seen the importance of understanding several factors, without knowing anything about waterfront, we coud have thought that old houses are still competitive and may have higher mean price

In [228]:
df_grouped = df[['new','bedrooms', 'bathrooms', 'floors', 'sqft_lot', 'sqft_living', 'grade']].groupby('new').agg({'bedrooms':'median', 'bathrooms':'median', 'floors':'median', 'sqft_living':'mean', 'sqft_lot':'mean', 'grade':'median'})
df_grouped

bedrooms  bathrooms  floors  sqft_living      sqft_lot  grade
new                                                                 
False       3.0        2.0     1.0  1969.397116  15916.799395    7.0
True        3.0        2.5     2.0  2471.244799  12238.947888    8.0

so in addition to being newer, houses built in 2000 onwards generally have more floors and larger sqft_living (which may be a trend in new houses construction) consequently having higher grade and higher price than older houses  

### Renovation

In [229]:
df['renovated'] = df['yr_renovated'] != 0
df_grouped = df[['price','renovated','id']].groupby('renovated').agg({'price':'mean','id':'count'})
df_grouped.rename(columns={'id':'count'},inplace=True)
df_grouped


price  count
renovated                      
False      530447.958597  20699
True       760628.777899    914

renovated houses have higher mean price than non-renovated, because renovation is so important especially that most houses are 30+ years old as we've seen

In [230]:
df.groupby('new')['renovated'].value_counts()

new    renovated
False  False        15941
       True           913
True   False         4758
       True             1
Name: renovated, dtype: int64

none of the houses built in 2000 onwards are renovated (except 1), because they don't need that actually, wait a minute, when does a house need renovation?

In [231]:
temp = df.loc[df.renovated, ['price','yr_built','yr_renovated']]
temp['difference'] = temp.yr_renovated - temp.yr_built
temp.difference.mean().round(2)

56.3

Renovated houses are renovated in average after 56.3 years from being built, we'll use that as age where the house needs renovation then  
So now we can clearly see that new houses built after 2000 still have a long life before needing renovation

In [232]:
temp = df.groupby('waterfront')['renovated'].value_counts(normalize=True)*100
temp

waterfront  renovated
0           False        95.934732
            True          4.065268
1           False        74.233129
            True         25.766871
Name: renovated, dtype: float64

only 5% of houses with no waterfront are renovated, while 25% of houses with waterfront are.  
  
Having a waterfront adds a lot of value to the house, that's why houses that have waterfront are more often renovated to retain their value

In [233]:
df['age'] = df.apply(lambda x:int(x.date[:4]) - (x.yr_built if not x.renovated else x.yr_renovated), axis=1)
df[['age','grade']].groupby('grade').mean().round()

age
grade      
1      51.0
3      69.0
4      73.0
5      78.0
6      69.0
7      49.0
8      31.0
9      23.0
10     22.0
11     21.0
12     21.0
13     23.0

this illustrates the effect of the age of the house on its grade, as low graded house are older and 9+ graded houses are less than 30 years old

### Decades

In [234]:
# create data frame that groups houses according to yr_built in decades
bins = [1900 + 10*i for i in range(1, 13)]
labels = [str(bins[i])+'-'+str(bins[i+1]) for i in range(len(bins)-1)]

# dictionary of dictionaries
data = {}

# for each decade, calculate aggregates, count, mean price, renovated percentage, and add them to dictionary
for i in range(11):
    curr = df.loc[(df.yr_built < bins[i+1]) & (df.yr_built > bins[i]), ['price','renovated', 'floors', 'bathrooms','sqft_living','sqft_lot']]
    data[labels[i]] = {'count':curr.price.count(), 'renovated':(curr.renovated.sum()/curr.renovated.count()).round(4)*100, 'mean_price':curr.price.mean().round(-3), 'floors':curr.floors.median(), 'bathrooms':curr.bathrooms.median(),'sqft_living':curr.sqft_living.mean().round(-2),'sqft_lot':curr.sqft_lot.mean().round(-2)}

df_decades = pd.DataFrame(data).transpose()
df_decades

count  renovated  mean_price  floors  bathrooms  sqft_living  \
1910-1920   672.0      14.73    570000.0     1.5       1.75       1700.0   
1920-1930  1094.0       9.96    607000.0     1.5       1.75       1800.0   
1930-1940   440.0      14.77    603000.0     1.0       1.75       1800.0   
1940-1950  1608.0       8.77    434000.0     1.0       1.00       1500.0   
1950-1960  2202.0       5.50    480000.0     1.0       1.75       1800.0   
1960-1970  2412.0       3.69    457000.0     1.0       1.75       1900.0   
1970-1980  2153.0       2.09    498000.0     1.0       2.25       2100.0   
1980-1990  2042.0       1.03    532000.0     2.0       2.25       2200.0   
1990-2000  1918.0       0.26    587000.0     2.0       2.50       2500.0   
2000-2010  3300.0       0.03    608000.0     2.0       2.50       2400.0   
2010-2020  1098.0       0.00    645000.0     2.0       2.50       2500.0   

           sqft_lot  
1910-1920    9900.0  
1920-1930    8000.0  
1930-1940   16500.0  
1940-1950   10600.0  
1950-1960   12000.0  
1960-1970   13800.0  
1970-1980   19600.0  
1980-1990   21700.0  
1990-2000   24400.0  
2000-2010   13600.0  
2010-2020    6300.0

This shows the distribution of houses based on the decade when they were built.  
We can find 2 insights from this grouping:  
* Most of the houses were built between 1950 and 1990
* newer houses tend to have more floors, more bathrooms  
* This grouping shows that most of renovated houses are the older ones that need it so much, but that also is not big percentage being renovated

In [235]:
print(f"{df.loc[(df.renovated == False) & ((2015 - df.yr_built) > 56)].shape[0]} houses are older than 56 years but not renovated")

6334 houses are older than 56 years but not renovated


6334 houses are older than 56 years but not renovated !  
this shows the great gap there, and the big opportunity to leverage the houses' mean price by renovating houses  
if you have you have a building company this is a good chance to renovate houses (especially with waterfronts) and then sell them with higher price

In [236]:
df['decade_built'] = df['decade_built'] = df.apply(lambda x: str((x.yr_built//10)*10)+'-'+str((x.yr_built//10+1)*10), axis=1)
df_decades = df_decades.reset_index().rename(columns={'index':'decade_built'})

merged = pd.merge(df, df_decades[['mean_price','decade_built']], on='decade_built', how='left')
merged.rename(columns={'mean_price':'dc_price'}, inplace=True)
merged['special'] = merged.price > merged.dc_price
merged['renovated'] = merged.yr_renovated != 0

merged

id             date     price  bedrooms  bathrooms  \
0      7129300520  20141013T000000  221900.0         3       1.00   
1      6414100192  20141209T000000  538000.0         3       2.25   
2      5631500400  20150225T000000  180000.0         2       1.00   
3      2487200875  20141209T000000  604000.0         4       3.00   
4      1954400510  20150218T000000  510000.0         3       2.00   
...           ...              ...       ...       ...        ...   
21608   263000018  20140521T000000  360000.0         3       2.50   
21609  6600060120  20150223T000000  400000.0         4       2.50   
21610  1523300141  20140623T000000  402101.0         2       0.75   
21611   291310100  20150116T000000  400000.0         3       2.50   
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...      lat     long  \
0             1180      5650     1.0           0     0  ...  47.5112 -122.257   
1             2570      7242     2.0           0     0  ...  47.7210 -122.319   
2              770     10000     1.0           0     0  ...  47.7379 -122.233   
3             1960      5000     1.0           0     0  ...  47.5208 -122.393   
4             1680      8080     1.0           0     0  ...  47.6168 -122.045   
...            ...       ...     ...         ...   ...  ...      ...      ...   
21608         1530      1131     3.0           0     0  ...  47.6993 -122.346   
21609         2310      5813     2.0           0     0  ...  47.5107 -122.362   
21610         1020      1350     2.0           0     0  ...  47.5944 -122.299   
21611         1600      2388     2.0           0     0  ...  47.5345 -122.069   
21612         1020      1076     2.0           0     0  ...  47.5941 -122.299   

       sqft_living15  sqft_lot15    new  renovated  age  decade_built  \
0               1340        5650  False      False   59     1950-1960   
1               1690        7639  False       True   23     1950-1960   
2               2720        8062  False      False   82     1930-1940   
3               1360        5000  False      False   49     1960-1970   
4               1800        7503  False      False   28     1980-1990   
...              ...         ...    ...        ...  ...           ...   
21608           1530        1509   True      False    5     2000-2010   
21609           1830        7200   True      False    1     2010-2020   
21610           1020        2007   True      False    5     2000-2010   
21611           1410        1287   True      False   11     2000-2010   
21612           1020        1357   True      False    6     2000-2010   

       dc_price  special  
0      480000.0    False  
1      480000.0     True  
2      603000.0    False  
3      457000.0     True  
4      532000.0    False  
...         ...      ...  
21608  608000.0    False  
21609  645000.0    False  
21610  608000.0    False  
21611  608000.0    False  
21612  608000.0    False  

[21613 rows x 27 columns]

In [237]:
merged.pivot_table(index='special', columns=['renovated','waterfront'],values='id', aggfunc='count').round()

renovated   False      True     
waterfront      0    1     0   1
special                         
False       13505   13   427   3
True         7073  108   445  39

this pivot table highlights the importance of renovation and having a waterfront, as those houses tend to be special in their decade more than others

### sqft

In [238]:
c = df['sqft_living'].corr(df['price'])
c2 = df['sqft_lot'].corr(df['price'])
print(f'corr living-price: {c.round(3)}\ncorr price-lot: {c2.round(3)}')

corr living-price: 0.702
corr price-lot: 0.09


There is a strong correlation between area and price (0.702), but surprisingly, the correlation between price and sqft_lot is weak, only 0.09
so as sqft_living increase, price increase

In [239]:
from math import ceil
p80 = float(np.percentile(df['sqft_lot'], 80))
print(f"80% of houses' areas lie below {ceil(p80)}, so far from the maximum {df['sqft_lot'].max()}")

80% of houses' areas lie below 12187, so far from the maximum 1651359


the distribution of sqft_lot is not uniform at all.  
so is the correlation better if we exclude all big houses (left 10%) ?

In [240]:
df_sqft = df[['price','sqft_lot']]
min = df['sqft_lot'].min()
max = p80
rang = max - min
nbins = 8
bins = [round((min + i*(rang/(nbins-1))),-3)  for i in range(nbins)]
bins[-1] = max
df_sqft['label'] = pd.cut(df['sqft_lot'], bins=bins, labels=[f'{bins[i]}-{bins[i+1]}' for i in range(len(bins)-1)])

C:\Users\Admin\AppData\Local\Temp\ipykernel_6840\4256706996.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_sqft['label'] = pd.cut(df['sqft_lot'], bins=bins, labels=[f'{bins[i]}-{bins[i+1]}' for i in range(len(bins)-1)])


In [241]:
df_sqft_grouped = df_sqft.groupby('label').agg({'price':'mean','sqft_lot':'count'}).reset_index()
df_sqft_grouped.rename(columns={'sqft_lot':'count', 'price':'mean price'}, inplace=True)
df_sqft_grouped

label     mean price  count
0               1000.0-2000.0  450308.099206   1008
1               2000.0-4000.0  534600.984103   2013
2               4000.0-6000.0  533032.880203   4132
3               6000.0-7000.0  487976.095788   1733
4               7000.0-9000.0  457860.871159   4719
5              9000.0-11000.0  512854.507446   2686
6  11000.0-12186.600000000002  564809.205290    794

the difference between houses areas (sqft_lot) at this level is not so decisive for price as other factors. It's not that sqft_lot doesn't matter, but it's that a higher sqft_lot does not necessarily mean higher price as it still depends heavily on other factors.

Note that this does not contradicts with the strong correlation btw sqft_living and price, but it shows that people -naturally- are more interested in the living area the actual area of the house more than total land area

In [242]:
df_decades[['sqft_living','sqft_lot']]

sqft_living  sqft_lot
0        1700.0    9900.0
1        1800.0    8000.0
2        1800.0   16500.0
3        1500.0   10600.0
4        1800.0   12000.0
5        1900.0   13800.0
6        2100.0   19600.0
7        2200.0   21700.0
8        2500.0   24400.0
9        2400.0   13600.0
10       2500.0    6300.0

From the decades dataframe we've made, we can see that this is the trend actually, newer houses have larger sqft_living, and smaller sqft_lot than that in 80s and 90s houses 

### Bedrooms

In [243]:
temp = df[['id','price','grade','bedrooms']].groupby('bedrooms').agg({'id':'count', 'price':'mean','grade':'median'})
temp = temp.rename(columns={'id':'count'}).reset_index()
temp

bedrooms  count         price  grade
0          0     13  4.102231e+05    7.0
1          1    199  3.176580e+05    6.0
2          2   2760  4.013877e+05    7.0
3          3   9824  4.662766e+05    7.0
4          4   6882  6.355647e+05    8.0
5          5   1601  7.868741e+05    8.0
6          6    272  8.258535e+05    8.0
7          7     38  9.514478e+05    8.0
8          8     13  1.105077e+06    8.0
9          9      6  8.939998e+05    7.0
10        10      3  8.200000e+05    7.0
11        11      1  5.200000e+05    7.0
12        33      1  6.400000e+05    7.0

as numbre of bedrooms increase, price increase

### Bathrooms

In [244]:
temp = df[['id','price','grade','bathrooms']].groupby('bathrooms').agg({'id':'count', 'price':'mean','grade':'median'})
temp = temp.rename(columns={'id':'count'}).reset_index()
temp

bathrooms  count         price  grade
0        0.00     10  4.490950e+05    7.0
1        0.50      4  2.373750e+05    6.0
2        0.75     72  2.945209e+05    6.0
3        1.00   3852  3.470412e+05    7.0
4        1.25      9  6.217722e+05    8.0
5        1.50   1446  4.093457e+05    7.0
6        1.75   3048  4.549158e+05    7.0
7        2.00   1930  4.579050e+05    7.0
8        2.25   2047  5.337688e+05    8.0
9        2.50   5380  5.536618e+05    8.0
10       2.75   1185  6.603505e+05    8.0
11       3.00    753  7.086619e+05    8.0
12       3.25    589  9.707532e+05    9.0
13       3.50    731  9.324017e+05    9.0
14       3.75    155  1.198179e+06   10.0
15       4.00    136  1.268405e+06   10.0
16       4.25     79  1.526653e+06   10.0
17       4.50    100  1.334211e+06   10.0
18       4.75     23  2.022300e+06   11.0
19       5.00     21  1.674167e+06   11.0
20       5.25     13  1.817962e+06   11.0
21       5.50     10  2.522500e+06   11.0
22       5.75      4  2.492500e+06   12.0
23       6.00      6  2.948333e+06   12.0
24       6.25      2  3.095000e+06   12.0
25       6.50      2  1.710000e+06   11.5
26       6.75      2  2.735000e+06   11.5
27       7.50      1  4.500000e+05    7.0
28       7.75      1  6.890000e+06   13.0
29       8.00      2  4.990000e+06   12.5

having at least 2 bathrooms is a must for any good house

it seems there is a rule that good houses must have good bathrooms, as we can see that prices increase as number of bathrooms increase, as well as the grade given for the house

### Floors

In [245]:
df_grouped = df[['floors','sqft_lot','sqft_living','waterfront','price','id','grade']].groupby('floors').agg({'price':'mean','id':'count','grade':'median','sqft_living':'mean','sqft_lot':'mean', 'waterfront':'sum'})
df_grouped = df_grouped.rename(columns={'id':'count'}).reset_index()
df_grouped.waterfront = df_grouped.apply(lambda x: x.waterfront / x['count'] *100, axis=1)
df_grouped.rename(columns={'waterfront':'waterfront percentage'}, inplace=True)
df_grouped

floors         price  count  grade  sqft_living      sqft_lot  \
0     1.0  4.422196e+05  10680    7.0  1731.846161  14568.405431   
1     1.5  5.590449e+05   1910    7.0  1907.880628  16966.476440   
2     2.0  6.490515e+05   8241    8.0  2571.279092  16133.744448   
3     2.5  1.061021e+06    161    9.0  3145.428571  17347.291925   
4     3.0  5.826201e+05    613    8.0  1788.311582   4461.562806   
5     3.5  9.339375e+05      8    8.0  2518.000000   3038.750000   

   waterfront percentage  
0               0.533708  
1               1.099476  
2               0.910084  
3               1.242236  
4               1.305057  
5               0.000000

houses with 2 floors or more tend to have higher grade and larger sqft_living, and it is important to note that waterfront is independent from number of floors

### Location

In [246]:
long = df.long
r = long.max() - long.min()
limits = [long.min() + (r/4)*i for i in range(4)]

labels = ['western','middle','eastern']
df['locationlong'] = df.apply(lambda x: labels[int(x.long > limits[1]) +int(x.long>limits[2])], axis=1)

t = df[['id','locationlong','decade_built','price','grade','waterfront']].groupby('locationlong').agg({'price':'mean','grade':'median','decade_built':lambda x:pd.Series.mode(x)[0],'id':'count','waterfront':'sum'}).reset_index()
t.price = t.price.round(-3)
t.rename(columns={'id':'count','price':'mean price','decade_built':'decade_built mode','grade':'grade median','waterfront':'#watefront'},inplace=True)
t

locationlong  mean price  grade median decade_built mode  count  #watefront
0      eastern    477000.0           7.0         2000-2010    719           0
1       middle    562000.0           8.0         2000-2010   9524          47
2      western    526000.0           7.0         1950-1960  11370         116

by adding a column that divides the area into 3 parts by longitude, its clear that the western part of this area includes a lot more houses   
  
most houses in western part are old, while new houses tend to be built in middle part

eastern part seems to have a small number of houses with less mean price compared to other 2, this may be explained by the fact that none have waterfront

most houses with water front are in the western part, and some are in middle, which may indicate that there is a lake or sea west the area where all houses lie

In [247]:
latt = df.lat
r = latt.max() - latt.min()
limits = [latt.min() + (r/4)*i for i in range(4)]

labels = ['southern','middle','northern']
df['location'] = df.apply(lambda x: labels[int(x.lat > limits[1]) +int(x.lat>limits[2])], axis=1)

t = df[['id','location','decade_built','price','grade','waterfront']].groupby('location').agg({'price':'mean','grade':'median','decade_built':lambda x:pd.Series.mode(x)[0],'id':'count','waterfront':'sum'}).reset_index()
t.price = t.price.round(-3)
t.rename(columns={'id':'count','price':'mean price','decade_built':'decade_built mode','grade':'grade median','waterfront':'#watefront'},inplace=True)
t

location  mean price  grade median decade_built mode  count  #watefront
0    middle    337000.0           7.0         1960-1970   4183          48
1  northern    609000.0           8.0         2000-2010  16313         115
2  southern    293000.0           7.0         1990-2000   1117           0

the 3 parts (divided by latitude) show the superiority of northern part over others in terms of number of houses and prices of those houses, and even number of houses with waterfront

for the waterfront, these percentages show that this waterfront is north western this area we are studying

#### Location and Price Level

In [248]:
min = df.price.min()
max = np.percentile(df.price, 90)
diff = max - min
nbins = 5
bins = [(min + i*(diff/(nbins-1))).round(-3) for i in range(nbins)]
bins.append(df.price.max())
nbins += 1
labels = ['very cheap', 'cheap','mid','expensive','very expensive']
labels = [labels[i] +': ' + str(bins[i])+'-'+str(bins[i+1]) for i in range(nbins-1)]

df['price_level'] = pd.cut(df['price'], bins=bins, labels=labels)

to avoid having awkward distribution and irrelevant groupby, I customized the bins so the first 4 are equal intervals but last one includes most expensive 10% whatever interval they take

In [249]:
df[['price_level','grade','bedrooms','bathrooms','floors','id']].groupby('price_level').agg({'id':'count','grade':'median','bedrooms':'median','bathrooms':'median','floors':'median'}).rename(columns={'id':'count'})

count  grade  bedrooms  bathrooms  floors
price_level                                                                  
very cheap: 75000.0-278000.0         3531    7.0       3.0       1.50     1.0
cheap: 278000.0-481000.0             8347    7.0       3.0       2.00     1.0
mid: 481000.0-684000.0               5081    8.0       3.0       2.25     1.5
expensive: 684000.0-887000.0         2492    8.0       4.0       2.50     2.0
very expensive: 887000.0-7700000.0   2161    9.0       4.0       3.00     2.0

ensuring the previous results, the more a house has bedrooms, bathrooms, floors, the higher price it has

In [250]:
df[['price_level', 'location']].groupby('location')['price_level'].value_counts().round().reset_index(name='count').sort_values(['location', 'price_level'], ascending=True)

location                         price_level  count
1     middle        very cheap: 75000.0-278000.0   1604
0     middle            cheap: 278000.0-481000.0   2124
2     middle              mid: 481000.0-684000.0    324
3     middle        expensive: 684000.0-887000.0     89
4     middle  very expensive: 887000.0-7700000.0     42
9   northern        very cheap: 75000.0-278000.0   1301
5   northern            cheap: 278000.0-481000.0   5806
6   northern              mid: 481000.0-684000.0   4699
7   northern        expensive: 684000.0-887000.0   2391
8   northern  very expensive: 887000.0-7700000.0   2116
10  southern        very cheap: 75000.0-278000.0    626
11  southern            cheap: 278000.0-481000.0    417
12  southern              mid: 481000.0-684000.0     58
13  southern        expensive: 684000.0-887000.0     12
14  southern  very expensive: 887000.0-7700000.0      3

it is clear that southern and middle parts are for cheap and lower middle houses, while northern part contains a lot more houses, and we can link that to the fact that a lot of houses in northern part are post 2000, while waterfront also exists west north the area

flipping the sides may help clear out some observations

In [251]:
df[['price_level','location']].groupby('price_level')['location'].value_counts(normalize=True).mul(100).round().reset_index(name='percentage within price_level')

price_level  location  \
0         very cheap: 75000.0-278000.0    middle   
1         very cheap: 75000.0-278000.0  northern   
2         very cheap: 75000.0-278000.0  southern   
3             cheap: 278000.0-481000.0  northern   
4             cheap: 278000.0-481000.0    middle   
5             cheap: 278000.0-481000.0  southern   
6               mid: 481000.0-684000.0  northern   
7               mid: 481000.0-684000.0    middle   
8               mid: 481000.0-684000.0  southern   
9         expensive: 684000.0-887000.0  northern   
10        expensive: 684000.0-887000.0    middle   
11        expensive: 684000.0-887000.0  southern   
12  very expensive: 887000.0-7700000.0  northern   
13  very expensive: 887000.0-7700000.0    middle   
14  very expensive: 887000.0-7700000.0  southern   

    percentage within price_level  
0                            45.0  
1                            37.0  
2                            18.0  
3                            70.0  
4                            25.0  
5                             5.0  
6                            92.0  
7                             6.0  
8                             1.0  
9                            96.0  
10                            4.0  
11                            0.0  
12                           98.0  
13                            2.0  
14                            0.0

so most -actually all- expensive and very expensive houses are in northern part, while most of very cheap houses are in mid and southern part

#### Most expensive area

find distance method:

In [252]:
from math import sin, cos, acos, radians
def find_distance(h1, h2):
    try:
        h1 = df[df.id == h1].iloc[0]
    except:
        print(f'first ID "{h1}" is not valid')
        return
    try:
        h2 = df[df.id == h2].iloc[0]
    except:
        print(f'second ID "{h2}" is not valid')
        return
    
    lat1, lon1, lat2, lon2 = map(lambda x:radians(x), [h1.lat, h1.long, h2.lat, h2.long])
    
    return acos(sin(lat1)*sin(lat2)+cos(lat1)*cos(lat2)*cos(lon2 - lon1))*6371

locating a center for most expensive houses area:

In [253]:
richest = df.sort_values(by='price', ascending=False).head(100)

richest.sort_values(['lat','long'], inplace=True)
richest = richest.reset_index().drop('index',axis=1)

center_house = richest.iloc[richest.shape[0]//2]
center = center_house['id']

for i in range(richest.shape[0]):
    richest.loc[i, 'dist_center'] = find_distance(richest.loc[i, 'id'], center)
radius = richest.dist_center.median()

print(f"then, most expensive houses are located in the circle of center (lat:{center_house.lat}, long:{center_house.long}) and radius {round(radius,2)}")

then, most expensive houses are located in the circle of center (lat:47.6263, long:-122.314) and radius 7.51


### Most expensive houses

In [254]:
p99 = np.percentile(df.price, 99)
richest = df.loc[df.price > p99]
richest

id             date      price  bedrooms  bathrooms  \
21     2524049179  20140826T000000  2000000.0         3       2.75   
153    7855801670  20150401T000000  2250000.0         4       3.25   
246    2025069065  20140929T000000  2400000.0         4       2.50   
269    7960900060  20150504T000000  2900000.0         4       3.25   
282    7424700045  20150513T000000  2050000.0         5       3.00   
...           ...              ...        ...       ...        ...   
21368  9533100285  20140630T000000  2070000.0         4       3.75   
21467  2311400056  20141201T000000  1990000.0         5       3.50   
21506  2524069097  20140509T000000  2240000.0         5       6.50   
21530  8964800330  20150407T000000  3000000.0         4       3.75   
21576  9253900271  20150107T000000  3570000.0         5       4.50   

       sqft_living  sqft_lot  floors  waterfront  view  ...     long  \
21            3050     44867     1.0           0     4  ... -122.233   
153           5180     19850     2.0           0     3  ... -122.162   
246           3650      8354     1.0           1     4  ... -122.072   
269           5050     20100     1.5           0     2  ... -122.223   
282           3830      8480     2.0           0     1  ... -122.287   
...            ...       ...     ...         ...   ...  ...      ...   
21368         4350      7965     2.0           0     0  ... -122.205   
21467         5230      8960     2.0           0     0  ... -122.201   
21506         7270    130017     2.0           0     0  ... -121.982   
21530         5090     14823     1.0           0     0  ... -122.207   
21576         4850     10584     2.0           1     4  ... -122.110   

       sqft_living15  sqft_lot15    new  renovated  age  decade_built  \
21              4110       20336  False      False   46     1960-1970   
153             3160        9750   True      False    9     2000-2010   
246             3120       18841   True      False   14     2000-2010   
269             3890       20060  False       True    7     1980-1990   
282             3050        7556  False       True   21     1900-1910   
...              ...         ...    ...        ...  ...           ...   
21368           2190        8557   True      False    1     2010-2020   
21467           2310        9603   True      False    0     2010-2020   
21506           1800       44890   True      False    4     2010-2020   
21530           3030       12752   True      False    2     2010-2020   
21576           3470       18270   True      False    8     2000-2010   

       locationlong  location                         price_level  
21          western  northern  very expensive: 887000.0-7700000.0  
153          middle  northern  very expensive: 887000.0-7700000.0  
246          middle  northern  very expensive: 887000.0-7700000.0  
269         western  northern  very expensive: 887000.0-7700000.0  
282         western  northern  very expensive: 887000.0-7700000.0  
...             ...       ...                                 ...  
21368        middle  northern  very expensive: 887000.0-7700000.0  
21467        middle  northern  very expensive: 887000.0-7700000.0  
21506        middle  northern  very expensive: 887000.0-7700000.0  
21530        middle  northern  very expensive: 887000.0-7700000.0  
21576        middle  northern  very expensive: 887000.0-7700000.0  

[217 rows x 28 columns]

In [255]:
print(f"the 1% most expensive houses include {((richest.waterfront.sum()/df.waterfront.sum())*100).round(2)}% of houses that have waterfront!")

the 1% most expensive houses include 30.06% of houses that have waterfront!


In [256]:
richest.describe()

id         price    bedrooms   bathrooms   sqft_living  \
count  2.170000e+02  2.170000e+02  217.000000  217.000000    217.000000   
mean   3.949840e+09  2.707558e+06    4.350230    3.820276   4979.534562   
std    2.737953e+09  8.366517e+05    1.016946    1.066082   1539.418267   
min    4.610050e+07  1.970000e+06    2.000000    1.750000   2610.000000   
25%    1.732800e+09  2.210000e+06    4.000000    3.250000   4040.000000   
50%    3.377900e+09  2.450000e+06    4.000000    3.750000   4600.000000   
75%    6.065300e+09  2.950000e+06    5.000000    4.250000   5480.000000   
max    9.831200e+09  7.700000e+06    8.000000    8.000000  13540.000000   

            sqft_lot      floors  waterfront        view   condition  ...  \
count     217.000000  217.000000  217.000000  217.000000  217.000000  ...   
mean    26320.456221    1.868664    0.225806    2.013825    3.525346  ...   
std     72080.206418    0.478747    0.419079    1.692793    0.751882  ...   
min      3452.000000    1.000000    0.000000    0.000000    2.000000  ...   
25%      9520.000000    2.000000    0.000000    0.000000    3.000000  ...   
50%     15682.000000    2.000000    0.000000    2.000000    3.000000  ...   
75%     22257.000000    2.000000    0.000000    4.000000    4.000000  ...   
max    920423.000000    3.500000    1.000000    4.000000    5.000000  ...   

        sqft_above  sqft_basement     yr_built  yr_renovated       zipcode  \
count   217.000000     217.000000   217.000000    217.000000    217.000000   
mean   3975.359447    1004.175115  1972.571429    294.483871  98063.788018   
std    1300.256826     856.760083    35.328167    709.713401     55.137213   
min    1520.000000       0.000000  1902.000000      0.000000  98004.000000   
25%    3100.000000       0.000000  1940.000000      0.000000  98008.000000   
50%    3750.000000    1010.000000  1988.000000      0.000000  98040.000000   
75%    4500.000000    1560.000000  2003.000000      0.000000  98112.000000   
max    9410.000000    4820.000000  2015.000000   2013.000000  98199.000000   

              lat        long  sqft_living15     sqft_lot15         age  
count  217.000000  217.000000     217.000000     217.000000  217.000000  
mean    47.618655 -122.237682    3460.059908   19842.539171   34.308756  
std      0.049432    0.090758     766.283746   35218.387768   33.110527  
min     47.416900 -122.410000    1680.000000    2199.000000    0.000000  
25%     47.585000 -122.287000    2960.000000    8620.000000    9.000000  
50%     47.626300 -122.231000    3450.000000   13639.000000   20.000000  
75%     47.638700 -122.210000    3940.000000   19845.000000   54.000000  
max     47.761200 -121.756000    6210.000000  411962.000000  112.000000  

[8 rows x 21 columns]

Expensive Houses Characteristics:
 - 2+ floors
 - waterfront
 - 4-5 bedrooms
 - 4 bathrooms
 - 5000 sqft living
 - 3 minimum condition

### Cheapest houses

In [257]:
p01 = np.percentile(df.price, 1)

cheapest = df[df.price <= p01]
cheapest.describe()

id          price    bedrooms   bathrooms  sqft_living  \
count  2.170000e+02     217.000000  217.000000  217.000000   217.000000   
mean   4.195125e+09  128428.354839    2.382488    1.043779   960.894009   
std    2.779334e+09   20782.004753    0.773560    0.260955   258.513115   
min    7.200179e+06   75000.000000    0.000000    0.000000   290.000000   
25%    1.843200e+09  115000.000000    2.000000    1.000000   780.000000   
50%    3.512100e+09  133400.000000    2.000000    1.000000   910.000000   
75%    7.129800e+09  147500.000000    3.000000    1.000000  1130.000000   
max    9.543000e+09  153500.000000    5.000000    2.000000  1730.000000   

            sqft_lot      floors  waterfront   view   condition  ...  \
count     217.000000  217.000000       217.0  217.0  217.000000  ...   
mean    10082.820276    1.066820         0.0    0.0    3.161290  ...   
std     15593.771377    0.212798         0.0    0.0    0.755703  ...   
min      1500.000000    1.000000         0.0    0.0    1.000000  ...   
25%      5393.000000    1.000000         0.0    0.0    3.000000  ...   
50%      7500.000000    1.000000         0.0    0.0    3.000000  ...   
75%      9750.000000    1.000000         0.0    0.0    4.000000  ...   
max    212137.000000    2.000000         0.0    0.0    5.000000  ...   

        sqft_above  sqft_basement     yr_built  yr_renovated       zipcode  \
count   217.000000     217.000000   217.000000    217.000000    217.000000   
mean    943.198157      17.695853  1949.147465     54.857143  98092.654378   
std     254.054023      82.849929    17.895178    326.092584     67.457911   
min     290.000000       0.000000  1900.000000      0.000000  98001.000000   
25%     780.000000       0.000000  1942.000000      0.000000  98030.000000   
50%     900.000000       0.000000  1949.000000      0.000000  98092.000000   
75%    1090.000000       0.000000  1961.000000      0.000000  98166.000000   
max    1730.000000     620.000000  1983.000000   2009.000000  98198.000000   

              lat        long  sqft_living15     sqft_lot15         age  
count  217.000000  217.000000     217.000000     217.000000  217.000000  
mean    47.438656 -122.245843    1289.013825   10110.179724   64.036866  
std      0.106083    0.146673     321.666386   17496.415991   18.972410  
min     47.210700 -122.398000     740.000000    1848.000000    5.000000  
25%     47.358500 -122.336000    1090.000000    6000.000000   52.000000  
50%     47.469000 -122.286000    1260.000000    7752.000000   64.000000  
75%     47.503000 -122.200000    1460.000000    9600.000000   72.000000  
max     47.736900 -121.315000    3290.000000  217800.000000  115.000000  

[8 rows x 21 columns]

In [258]:
cheapest.loc[:, 'renovated'] = cheapest.yr_renovated != 0
print(f"Only {str((cheapest.renovated.sum()/cheapest.shape[0])*100)[:4]}% of cheapest houses are renovated")

Only 2.76% of cheapest houses are renovated


C:\Users\Admin\AppData\Local\Temp\ipykernel_6840\2374161140.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  cheapest.loc[:, 'renovated'] = cheapest.yr_renovated != 0


Cheapest Houses Characteristics:
 - 1 floor
 - no waterfront
 - 2 bedrooms
 - 1 bathrooms
 - sqft living < 1000
 - 40+ years old & not renovated